libgit/autogit_manager.go

```diff
@@ -225,10 +282,84 @@ func (am *AutogitManager) TlfHandleChange(
 	// Do nothing.
 }
 
+// GetBrowserForRepo returns the root FS for the specified repo and a
+// `Browser` for the branch and subdir.
+func (am *AutogitManager) GetBrowserForRepo(
+	ctx context.Context, gitFS *libfs.FS, repoName string,
+	branch plumbing.ReferenceName, subdir string) (*libfs.FS, *Browser, error) {
+	key := browserCacheKey{gitFS, repoName, branch, subdir}
+
+	am.browserLock.Lock()
+	defer am.browserLock.Unlock()
+	tmp, ok := am.browserCache.Get(key)
+	if ok {
+		b, ok := tmp.(browserCacheValue)
+		if !ok {
+			return nil, nil, errors.Errorf("Bad browser in cache: %T", tmp)
+		}
+		return b.repoFS, b.browser, nil
+	}
+
+	// It's kind of dumb to hold the browser lock through all of this,
+	// but it doesn't seem worthwhile to build the whole
+	// channel/notification system that would be needed to manage
+	// multiple concurrent requests for the same repo node.
+
+	am.log.CDebugf(ctx, "Making browser for repo=%s, branch=%s, subdir=%s",
+		repoName, branch, subdir)
+
+	rootKey := key
+	rootKey.subdir = ""
+
+	// Shortcut if the root of the repo already has a browser.
+	if subdir != "" {
+		tmp, ok = am.browserCache.Get(rootKey)
+		if ok {
+			rootB, ok := tmp.(browserCacheValue)
+			if ok {
+				b, err := rootB.browser.Chroot(subdir)
+				if err != nil {
+					return nil, nil, err
+				}
+				browser, ok := b.(*Browser)
+				if !ok {
+					return nil, nil, errors.Errorf("Bad browser type: %T", b)
+				}
+				am.browserCache.Add(
+					key, browserCacheValue{rootB.repoFS, browser})
+				return rootB.repoFS, browser, nil
+			}
```

Comment:
If `!ok`, might as well return `errors.Errorf("Bad browser in cache: %T", tmp)` to keep consistent?